File: microservice-architecture/examples/microservice-basics/go/src/http_inventory_client.go
package src

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
)

type HttpInventoryClient struct {
	BaseURL string
	Client  *http.Client
}

func NewHttpInventoryClient(baseURL string) *HttpInventoryClient {
	return &HttpInventoryClient{
		BaseURL: baseURL,
		Client:  &http.Client{},
	}
}

func (c *HttpInventoryClient) Reserve(sku string, quantity int) bool {
	endpoint := fmt.Sprintf("%s/reserve?sku=%s&quantity=%d", c.BaseURL, url.QueryEscape(sku), quantity)
	resp, err := c.Client.Get(endpoint)
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return false
	}

	return resp.StatusCode == http.StatusOK && string(body) == "OK"
}

File: microservice-architecture/examples/microservice-basics/ts/src/HttpInventoryClient.ts
import http from 'http'
import type { IncomingMessage } from 'http'

export class HttpInventoryClient {
  constructor(private readonly baseUrl: URL) {}

  reserve(sku: string, quantity: number): Promise<boolean> {
    return new Promise((resolve) => {
      const path = `/reserve?sku=${encodeURIComponent(sku)}&quantity=${quantity}`
      const options: http.RequestOptions = {
        hostname: this.baseUrl.hostname,
        port: this.baseUrl.port,
        path,
        method: 'GET',
      }

      const req = http.request(options, (res: IncomingMessage) => {
        let body = ''
        res.on('data', (chunk: Buffer) => {
          body += chunk.toString('utf-8')
        })
        res.on('end', () => {
          resolve(res.statusCode === 200 && body === 'OK')
        })
      })

      req.on('error', () => resolve(false))
      req.end()
    })
  }
}
